[view-transitions] Support chaining `:only-child` after pseudo-elements

nt1m opened the pull request:
#### d1418bf98d34738599cef6f5f77255c3174f6a8f
<pre>
[view-transitions] Support chaining `:only-child` after pseudo-elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=265223">https://bugs.webkit.org/show_bug.cgi?id=265223</a>
<a href="https://rdar.apple.com/118703154">rdar://118703154</a>

Reviewed by Antti Koivisto.

::view-transition-group(name):only-child matches a group with no siblings.

::view-transition-image-pair(name):only-child should always match, since there is only one image-pair.

::view-transition-old(name):only-child matches an old capture with no associated new element.

::view-transition-new(name):only-child matches a new element with no associated old capture.

* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/only-child-group-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/only-child-group.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/only-child-image-pair-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/only-child-image-pair.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/only-child-new-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/only-child-new.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/only-child-no-transition.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/only-child-old-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/only-child-old.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/parsing/pseudo-elements-valid-expected.txt:
* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::LocalContext::LocalContext):
(WebCore::SelectorChecker::match const):
(WebCore::SelectorChecker::matchHostPseudoClass const):
(WebCore::hasViewTransitionPseudoElement):
(WebCore::SelectorChecker::matchRecursively const):
(WebCore::SelectorChecker::checkOne const):
(WebCore::SelectorChecker::matchHasPseudoClass const):
(WebCore::SelectorChecker::checkViewTransitionPseudoClass const):
* Source/WebCore/css/SelectorChecker.h:
* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::isPseudoClassValidAfterPseudoElement):
* Source/WebCore/dom/ViewTransition.h:
(WebCore::OrderedNamedElementsMap::size const):
(WebCore::OrderedNamedElementsMap::find const):
* Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp:
(WebCore::createRendererIfNeeded):

Canonical link: <a href="https://commits.webkit.org/275841@main">https://commits.webkit.org/275841@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c62b10914eb3b9872b0c1093be0ee0f8445d69d4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43032 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22058 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45434 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45657 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39157 "Built successfully") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/9/builds/45338 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/25786 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19481 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/35581 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43605 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/19094 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/37063 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/16577 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/16685 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/38120 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1088 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/39216 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/38451 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/47190 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/17885 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/14736 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/42361 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/19484 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/37384 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/41018 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/19663 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5827 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/19114 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/SelectorChecker.cpp:
```diff
@@ -1503,6 +1525,44 @@ bool SelectorChecker::checkScrollbarPseudoClass(const CheckingContext& checkingC
+        return false;
+    }
+    case PseudoId::ViewTransition:
+        return false;
```
nit: move this to alpha order

nt1m replied:
fixed

mdubet commented on Source/WebCore/css/SelectorChecker.cpp:
```diff
@@ -177,7 +179,8 @@ SelectorChecker::SelectorChecker(Document& document)
 bool SelectorChecker::match(const CSSSelector& selector, const Element& element, CheckingContext& checkingContext) const
 {
-    LocalContext context(selector, element, checkingContext.resolvingMode == SelectorChecker::Mode::QueryingRules ? VisitedMatchType::Disabled : VisitedMatchType::Enabled, checkingContext.pseudoId);
+    auto pseudoElementIdentifier = checkingContext.pseudoId == PseudoId::None ? std::nullopt : std::optional(Style::PseudoElementIdentifier { checkingContext.pseudoId, checkingContext.pseudoElementNameArgument });
```
```
std::optional<PseudoElementIdentifier> pseudoElementIdentifier;
if (checkingContext.pseudoId != None)
   pseudoElementIdenfitier = Style::PseudoElementIdentifier { checkigContext.pseudoId, checkContext.pseudoElemet };
```

might be clearer ?

nt1m replied:
I like the one liner better, and the ternary should go away once I figure out how to switch `CheckingContext` and CSSJIT to use PseudoElementIdentifier

mdubet commented on Source/WebCore/css/SelectorChecker.cpp:
```diff
@@ -220,6 +223,20 @@ bool SelectorChecker::matchHostPseudoClass(const CSSSelector& selector, const El
     return true;
 }
 
+inline static bool hasViewTransitionPseudoElement(const PseudoIdSet& dynamicPseudoIdSet)
```
Why is it `inline` ?

nt1m replied:
I followed `hasScrollbarPseudoElement`, but no particular reason otherwise

mdubet commented on Source/WebCore/css/SelectorChecker.cpp:
```diff
@@ -1503,6 +1525,44 @@ bool SelectorChecker::checkScrollbarPseudoClass(const CheckingContext& checkingC
+        return false;
+    case PseudoId::ViewTransitionGroup: {
+        if (RefPtr activeViewTransition = element.document().activeViewTransition()) {
+            auto* capturedElement = activeViewTransition->namedElements().find(checkingContext.pseudoElementNameArgument);
```
It seems we don't need the `capturedElement` here ?

```
if (activeViewTransition->namedElements().find(checkingContext.pseudoElementNameArgument))
     return  activeViewTransition->namedElements().size() == 1
```

nt1m replied:
Sure but then the 2 liner becomes a 3 liner:
```
if (RefPtr activeViewTransition = element.document().activeViewTransition()) {
     if (activeViewTransition->namedElements().find(checkingContext.pseudoElementNameArgument))
          return  activeViewTransition->namedElements().size() == 1
     return false;
}
```

nt1m replied:
oh I think I see what you mean

mdubet commented on Source/WebCore/css/SelectorChecker.cpp:
```diff
@@ -1503,6 +1525,44 @@ bool SelectorChecker::checkScrollbarPseudoClass(const CheckingContext& checkingC
+        return true;
+    case PseudoId::ViewTransitionOld: {
+        if (RefPtr activeViewTransition = element.document().activeViewTransition()) {
+            auto* capturedElement = activeViewTransition->namedElements().find(checkingContext.pseudoElementNameArgument);
```
```
if (auto* capturedElement = activeViewTransition->namedElements().find(checkingContext.pseudoElementNameArgument))
     return  capturedElement->newElement;
```

mdubet commented on Source/WebCore/dom/ViewTransition.h:
```diff
@@ -52,7 +52,9 @@ enum class ViewTransitionPhase : uint8_t {
-    RefPtr<ImageBuffer> oldImage;
+    // std::nullopt represents an non-capturable element.
+    // nullptr represents an absent snapshot on an capturable element.
+    std::optional<RefPtr<ImageBuffer>> oldImage;
```
This kind of 3 states value would ideally have a proper variant

```
struct NonCapturable {}
struct AbsentSnapshot {}
std::variant<NonCapturable, AbsentSnapshot, Ref<ImageBuffer>>;
```

(not for this PR though)

nt1m replied:
I don't think the variant is a great idea, `std::optional` reads more like the spec, because lines like if "old image is not null" read exactly as `if (capturedElement->oldImage)`.

mdubet commented on Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp:
```diff
@@ -121,17 +121,25 @@ static RenderPtr<RenderBox> createRendererIfNeeded(RenderElement& documentElemen
-    auto newStyle = RenderStyle::clone(*style);
     RenderPtr<RenderBox> renderer;
     if (pseudoId == PseudoId::ViewTransitionOld || pseudoId == PseudoId::ViewTransitionNew) {
+        const auto* capturedElement = document->activeViewTransition()->namedElements().find(name);
```
It seems we can crash if `capturedElement` is `nullptr` here.

Either check for this, or ASSERT it's not if it's an impossible state.

mdubet commented on Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp:
```diff
@@ -121,17 +121,25 @@ static RenderPtr<RenderBox> createRendererIfNeeded(RenderElement& documentElemen
     RenderPtr<RenderBox> renderer;
     if (pseudoId == PseudoId::ViewTransitionOld || pseudoId == PseudoId::ViewTransitionNew) {
+        const auto* capturedElement = document->activeViewTransition()->namedElements().find(name);
+        if (pseudoId == PseudoId::ViewTransitionOld && !capturedElement->oldImage)
```
Here, we dereference `caputuredElement` without checking it's not null

mdubet commented on Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp:
```diff
@@ -121,17 +121,25 @@ static RenderPtr<RenderBox> createRendererIfNeeded(RenderElement& documentElemen
         renderer = WTFMove(rendererViewTransition);
-    } else
+    } else {
+        auto newStyle = RenderStyle::clone(*style);
```
Not sure there is an added value of having the `newStyle` variable to move it in the next line ? Maybe we should call `RenderStyle::clone` directly as the `createRenderer` argument.

anttijk commented on Source/WebCore/css/SelectorChecker.cpp:
```diff
@@ -220,6 +223,20 @@ bool SelectorChecker::matchHostPseudoClass(const CSSSelector& selector, const El
+    };
+    if (dynamicPseudoIdSet & functionalPseudoElementSet)
+        return true;
+    return false;
```
`return dynamicPseudoIdSet & functionalPseudoElementSet`

nt1m committed:
[view-transitions] Support chaining `:only-child` after pseudo-elements
https://bugs.webkit.org/show_bug.cgi?id=265223
rdar://118703154

Reviewed by Antti Koivisto.

::view-transition-group(name):only-child matches a group with no siblings.

::view-transition-image-pair(name):only-child should always match, since there is only one image-pair.

::view-transition-old(name):only-child matches an old capture with no associated new element.

::view-transition-new(name):only-child matches a new element with no associated old capture.

* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/only-child-group-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/only-child-group.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/only-child-image-pair-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/only-child-image-pair.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/only-child-new-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/only-child-new.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/only-child-no-transition.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/only-child-old-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/only-child-old.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/parsing/pseudo-elements-valid-expected.txt:
* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::LocalContext::LocalContext):
(WebCore::SelectorChecker::match const):
(WebCore::SelectorChecker::matchHostPseudoClass const):
(WebCore::hasViewTransitionPseudoElement):
(WebCore::SelectorChecker::matchRecursively const):
(WebCore::SelectorChecker::checkOne const):
(WebCore::SelectorChecker::matchHasPseudoClass const):
(WebCore::SelectorChecker::checkViewTransitionPseudoClass const):
* Source/WebCore/css/SelectorChecker.h:
* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::isPseudoClassValidAfterPseudoElement):
* Source/WebCore/dom/ViewTransition.h:
(WebCore::OrderedNamedElementsMap::size const):
(WebCore::OrderedNamedElementsMap::find const):
* Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp:
(WebCore::createRendererIfNeeded):

Canonical link: https://commits.webkit.org/275841@main
